### ISSUE TITLE
Why ETCD new leader will restart the TTL counting when leader swithed

### ISSUE THREAD

**Issue by jiapeish** [etcd-io/etcd#13294]
**Issue Overview**
I noticed that etcd will restart the TTL counting when leader swithed. 

**Steps to reproduce**
1. Start a etcd client and put a key with lease, like TTL = 10s;
2. When the key's remaining ttl becomes 4s, restart etcd leader node;
3. The key's remaining ttl becomes to 10s again.

Why not the TTL continue counting down after new leader is elected? And how can we modify the code to fix it?

**Comment by serathius** [comment]
This is problem in distributed systems, you cannot trust the time, each instance has different local time. A workaround for this is using time difference (TTL) instead of deadline. Etcd persists TTL and counts it from last leader change. If leader changes TTL will be preserved, but deadline will change. 

There is and experimental feature called lease checkpointing that should help with this (checkpoints TTL every 5 minutes).

**Comment by jiapeish** [comment]
> This is problem in distributed systems, you cannot trust the time, each instance has different local time. A workaround for this is using time difference (TTL) instead of deadline. Etcd persists TTL and counts it from last leader change. If leader changes TTL will be preserved, but deadline will change. Thus TTL for leases is not the exact, but the minimal time for lease to be present.
> 
> There is and experimental feature called lease checkpointing that should help with this (checkpoints TTL every 5 minutes). Proper solution would require Etcd members to agree on time, which would require more work. Contributions are welcomed!

Thank you @serathius, we'll turn the checkpointing feature on and have a test. 

As you mentioned, the TTL is checkpointed every 5 minutes, but I didn't find a loop/timer that do this checkpointing work periodically. How does etcd control the checkpoint period? Does it means that ETCD leader node will do this checkpoint and sends the remaining TTL to follower nodes every 5 minutes, so that follower will continue counting down?

**Comment by ardaguclu** [comment]
@jiapeish, This pushes requests to the heap per `checkpointinterval` https://github.com/etcd-io/etcd/blob/ab9563dc8e4d84d06080c765f9c6c0d50313f794/server/lease/lessor.go#L714 (`checkpointInterval` is customizable with 5 minutes default value and being used in here). 

For consumer side, This runs forever https://github.com/etcd-io/etcd/blob/ab9563dc8e4d84d06080c765f9c6c0d50313f794/server/lease/lessor.go#L588. Basically, this https://github.com/etcd-io/etcd/blob/ab9563dc8e4d84d06080c765f9c6c0d50313f794/server/lease/lessor.go#L754 updates remainingTTL.

If you wonder why this does not work when not enabling `experimental-enable-lease-checkpoint`. Because this https://github.com/etcd-io/etcd/blob/ab9563dc8e4d84d06080c765f9c6c0d50313f794/server/lease/lessor.go#L831 always returns ttl instead remainingTTL, since there is no scheduler to set this value.

@serathius I have one question. While I was investigating the issue, I realized this https://github.com/etcd-io/etcd/blob/ab9563dc8e4d84d06080c765f9c6c0d50313f794/server/lease/lessor.go#L755 condition. Is there any possibility that this can be true?

**Comment by busgo** [comment]
> > This is problem in distributed systems, you cannot trust the time, each instance has different local time. A workaround for this is using time difference (TTL) instead of deadline. Etcd persists TTL and counts it from last leader change. If leader changes TTL will be preserved, but deadline will change. Thus TTL for leases is not the exact, but the minimal time for lease to be present.
> > There is and experimental feature called lease checkpointing that should help with this (checkpoints TTL every 5 minutes). Proper solution would require Etcd members to agree on time, which would require more work. Contributions are welcomed!
> 
> Thank you @serathius, we'll turn the checkpointing feature on and have a test.
> 
> As you mentioned, the TTL is checkpointed every 5 minutes, but I didn't find a loop/timer that do this checkpointing work periodically. How does etcd control the checkpoint period? Does it means that ETCD leader node will do this checkpoint and sends the remaining TTL to follower nodes every 5 minutes, so that follower will continue counting down?

You can check whether there is a leader at regular time. Code like this

```go 

// start elect
func (n *PinkNode) electLoop() {
	ticker := time.Tick(time.Second * time.Duration(n.electionTTL))
	log.Printf("the pink node instance %s start elect loop....", n.id)
	for {
		select {
		case <-ticker:
			if n.electionState == ElectionReadyState {
				n.electionState = ElectionDoingState
				log.Printf("the pink node instance %s start try elect loop....", n.id)
				n.tryElect()
			}

		}
	}
}

// try  elect
func (n *PinkNode) tryElect() {

	defer func() {
		n.electionState = ElectionReadyState
	}()
	ctx, _ := context.WithTimeout(context.TODO(), time.Second*3)
	id, err := n.etcdCli.Leader(ctx, n.electionPath)
	if err == nil {
		log.Printf("the pink node instance %s has leader is %s", n.id, id)
		if id == n.id {
			n.NotifyState(protocol.Leader)
		} else {
			n.NotifyState(protocol.Follower)
		}
		return
	}
	n.NotifyState(protocol.Follower)
	log.Printf("the pink node instance %s find leader fail:%+v", n.id, err)
	if !errors.Is(err, concurrency.ErrElectionNoLeader) {
		log.Printf("the pink node %s get leader fail:%+v", n.id, err)
		return
	}

	log.Printf("the pink node instance %s start campaign  leader", n.id)
	err = n.etcdCli.Campaign(ctx, n.id, n.electionPath, n.electionTTL)
	if err == nil {
		log.Printf("the pink node instance %s campaign  leader success", n.id)
		n.NotifyState(protocol.Leader)
		return
	}

}

```

**Comment by jiapeish** [comment]
It's nice of you to explain this clearly @ardaguclu . It looks that I need to have a read of the lessor code...

**Comment by jiapeish** [comment]
Hey, thank you @busgo , I've just find the checkpoint feature and maybe it will help.

**Comment by serathius** [comment]
Looks like we lease checkpointing was not correctly implemented, proposed fix https://github.com/etcd-io/etcd/pull/13491

**Comment by jiapeish** [comment]
> Looks like we lease checkpointing was not correctly implemented, proposed fix #13491 
This fix is really import to me, I was wondering why the leasechecking not working after leader changed...

**Comment by serathius** [comment]
Closing as issue was resolved and fix will be released in v3.5.2
